File: ECONOROTA/api/test/optimizer.test.ts
import assert from 'node:assert/strict';
import { test } from 'node:test';

import { optimize, type MarketIn } from '../src/lib/optimizer.ts';

const mk = (id: string, fee: number, prices: Record<string, number>, opts: Partial<MarketIn> = {}): MarketIn => ({
  id, name: id, deliveryFeeCents: fee, minOrderCents: 0, etaMax: 30, distanceKm: 1,
  offers: new Map(Object.entries(prices).map(([k, p]) => [k, { productId: `${id}-${k}`, priceCents: p, stock: 10 }])),
  ...opts,
});
const want = (...keys: string[]) => keys.map((key) => ({ key, qty: 1 }));
const ten = 'abcdefghij'.split('');

test('um mercado quando a economia não paga a taxa extra', () => {
  const a = mk('A', 500, Object.fromEntries(ten.map((k) => [k, 100])));
  const b = mk('B', 500, Object.fromEntries(ten.map((k) => [k, 99])));
  const r = optimize([a, b], want(...ten));
  assert.deepEqual(r.cheapest!.marketIds, ['B']);
  assert.equal(r.cheapest!.totalCents, 990 + 500);
});

test('dois mercados quando compensa, cada um com pelo menos 5 itens', () => {
  const a = mk('A', 300, Object.fromEntries(ten.map((k, i) => [k, i < 5 ? 100 : 300])));
  const b = mk('B', 300, Object.fromEntries(ten.map((k, i) => [k, i < 5 ? 300 : 100])));
  const r = optimize([a, b], want(...ten));
  assert.equal(r.cheapest!.marketIds.length, 2);
  assert.equal(r.cheapest!.totalCents, 1000 + 600);
  for (const m of r.cheapest!.marketIds) assert.ok(r.cheapest!.lines.filter((l) => l.marketId === m).length >= 5);
});

test('nunca divide com menos de 5 itens por mercado', () => {
  // B é mais barato em só 3 itens: dividir violaria a regra, então fica tudo em A.
  const a = mk('A', 300, Object.fromEntries(ten.map((k) => [k, 200])));
  const b = mk('B', 300, Object.fromEntries(ten.map((k, i) => [k, i < 3 ? 10 : 500])));
  const r = optimize([a, b], want(...ten));
  assert.deepEqual(r.cheapest!.marketIds, ['A']);
});

test('rebalanceia itens para atingir o mínimo quando ainda compensa', () => {
  // B muito mais barato em 4 itens; o 5º item move com pequeno acréscimo e a divisão ainda vale a pena.
  const a = mk('A', 100, Object.fromEntries(ten.map((k) => [k, 200])));
  const b = mk('B', 100, Object.fromEntries(ten.map((k, i) => [k, i < 4 ? 10 : 260])));
  const r = optimize([a, b], want(...ten));
  assert.equal(r.cheapest!.marketIds.length, 2);
  assert.equal(r.cheapest!.lines.filter((l) => l.marketId === 'B').length, 5);
});

test('no máximo 3 mercados', () => {
  const keys = Array.from({ length: 20 }, (_, i) => `k${i}`);
  const markets = ['A', 'B', 'C', 'D'].map((id, mi) =>
    mk(id, 0, Object.fromEntries(keys.map((k, i) => [k, Math.floor(i / 5) === mi ? 1 : 1000]))),
  );
  const r = optimize(markets, want(...keys));
  assert.equal(r.cheapest!.marketIds.length, 3);
});

test('respeita estoque e informa itens faltando', () => {
  const a = mk('A', 0, { x: 100, y: 100 });
  a.offers.get('y')!.stock = 0;
  const r = optimize([a], [{ key: 'x', qty: 1 }, { key: 'y', qty: 1 }, { key: 'z', qty: 1 }]);
  assert.deepEqual(r.cheapest!.missing.sort(), ['y', 'z']);
});

test('respeita pedido mínimo do mercado', () => {
  const a = mk('A', 0, { x: 100 }, { minOrderCents: 1000 });
  const b = mk('B', 0, { x: 150 });
  const r = optimize([a, b], want('x'));
  assert.deepEqual(r.cheapest!.marketIds, ['B']);
});

test('rota: visita os mercados na ordem mais curta até a casa', async () => {
  const { bestRoute } = await import('../src/lib/optimizer.ts');
  // Linha reta: casa em x=0; mercados em x=1, 3, 2 → melhor ordem é 3, 2, 1.
  const d = (a: { lat: number; lng: number }, b: { lat: number; lng: number }) => Math.abs(a.lat - b.lat);
  const r = bestRoute([{ id: 'A', lat: 1, lng: 0 }, { id: 'B', lat: 3, lng: 0 }, { id: 'C', lat: 2, lng: 0 }], { lat: 0, lng: 0 }, d)!;
  assert.deepEqual(r.order, ['B', 'C', 'A']);
  assert.equal(r.total_km, 3);
  assert.deepEqual(r.legs_km, [1, 1, 1]);
});

File: ECONOROTA/app/test/optimizer_test.dart
import 'package:econorota/core/compare/compare_key.dart';
import 'package:econorota/core/compare/optimizer.dart';
import 'package:econorota/core/compare/route.dart';
import 'package:econorota/core/compare/rules.dart';
import 'package:flutter_test/flutter_test.dart';

MarketOffers mk(String id, int fee, Map<String, int> prices, {int minOrder = 0}) => MarketOffers(
  id: id,
  deliveryFeeCents: fee,
  minOrderCents: minOrder,
  offers: {for (final e in prices.entries) e.key: Offer(productId: '$id-${e.key}', priceCents: e.value, stock: 10)},
);

/// Testes das regras do motor com a taxa de cada mercado somada (isola a regra da entrega única).
({Plan? cheapest, Plan? single}) optimizeSum(List<MarketOffers> m, List<Want> w) =>
    optimize(m, w, fee: (ms) => ms.fold(0, (s, x) => s + x.deliveryFeeCents));

void main() {
  final ten = 'abcdefghij'.split('');
  List<Want> want(List<String> keys) => [for (final k in keys) Want(k, 1)];

  test('chave de comparação sem acento', () {
    expect(compareKey('Óleo de Soja', '900 ml'), 'oleo de soja|900ml');
    expect(compareKey('Feijão Carioca', '1 kg'), 'feijao carioca|1kg');
  });

  test('um mercado quando a divisão não compensa a taxa', () {
    final r = optimizeSum([
      mk('A', 500, {for (final k in ten) k: 100}),
      mk('B', 500, {for (final k in ten) k: 99}),
    ], want(ten));
    expect(r.cheapest!.marketIds, ['B']);
  });

  test('divide em 2 com no mínimo 5 itens por mercado', () {
    final a = mk('A', 300, {for (final (i, k) in ten.indexed) k: i < 5 ? 100 : 300});
    final b = mk('B', 300, {for (final (i, k) in ten.indexed) k: i < 5 ? 300 : 100});
    final r = optimizeSum([a, b], want(ten));
    expect(r.cheapest!.marketIds.length, 2);
    for (final m in r.cheapest!.marketIds) {
      expect(r.cheapest!.linesOf(m).length, greaterThanOrEqualTo(minItemsPerMarket));
    }
    expect(r.cheapest!.totalCents, 1600);
  });

  test('nunca divide com menos de 5 itens em um mercado', () {
    final a = mk('A', 300, {for (final k in ten) k: 200});
    final b = mk('B', 300, {for (final (i, k) in ten.indexed) k: i < 3 ? 10 : 500});
    expect(optimizeSum([a, b], want(ten)).cheapest!.marketIds, ['A']);
  });

  test('no máximo 3 mercados', () {
    final keys = [for (var i = 0; i < 20; i++) 'k$i'];
    final markets = [
      for (final (mi, id) in ['A', 'B', 'C', 'D'].indexed)
        mk(id, 0, {for (final (i, k) in keys.indexed) k: i ~/ 5 == mi ? 1 : 1000}),
    ];
    expect(optimizeSum(markets, want(keys)).cheapest!.marketIds.length, 3);
  });

  test('pedido mínimo e itens faltando', () {
    final r = optimizeSum([
      mk('A', 0, {'x': 100}, minOrder: 1000),
      mk('B', 0, {'x': 150}),
    ], want(['x', 'y']));
    expect(r.cheapest!.marketIds, ['B']);
    expect(r.cheapest!.missing, ['y']);
  });

  test('entrega única: base + adicional por mercado extra', () {
    expect(deliveryFeeFor(1), 790);
    expect(deliveryFeeFor(3), 1390);
    final a = mk('A', 0, {for (final (i, k) in ten.indexed) k: i < 5 ? 100 : 400});
    final b = mk('B', 0, {for (final (i, k) in ten.indexed) k: i < 5 ? 400 : 100});
    final r = optimize([a, b], want(ten));
    expect(r.cheapest!.marketIds.length, 2);
    expect(r.cheapest!.deliveryCents, 1090);
  });

  test('3 mercados: divide só quando cada um tem 5 itens e a economia paga as entregas extras', () {
    final keys = [for (var i = 0; i < 15; i++) 'k$i'];
    MarketOffers cheapIn(String id, int from) =>
        mk(id, 0, {for (final (i, k) in keys.indexed) k: i ~/ 5 == from ? 100 : 500});
    final r = optimize([cheapIn('A', 0), cheapIn('B', 1), cheapIn('C', 2)], [for (final k in keys) Want(k, 1)]);
    expect(r.cheapest!.marketIds.length, 3);
    expect(r.cheapest!.deliveryCents, deliveryFeeFor(3));
    for (final m in r.cheapest!.marketIds) {
      expect(r.cheapest!.linesOf(m).length, 5);
    }
    expect(r.cheapest!.totalCents, 1500 + 1390);
  });

  test('distância (Haversine) e melhor ordem de coleta até a casa', () {
    const paulista = (lat: -23.5614, lng: -46.6559);
    const se = (lat: -23.5503, lng: -46.6340);
    expect(distanceKm(paulista, se), closeTo(2.6, 0.2));
    const home = (lat: -23.5700, lng: -46.6450);
    final r = bestRoute({'longe': (lat: -23.5400, lng: -46.6300), 'perto': (lat: -23.5650, lng: -46.6480)}, home)!;
    expect(r.order, ['longe', 'perto']); // passa no mais distante primeiro e termina perto de casa
    expect(r.legsKm.length, 2);
    expect(r.minutes, greaterThan(0));
  });
}
